### DISCUSSION TITLE
livenessProbe failed on proxy container - strange behavior

### DISCUSSION THREAD

**Discussion by baznikin** [Kong/kong#10720] [Help]
Hello!

I am fighting with with issue for some time already with no success. We run Kong on Kubernetes, it is installed with Helm chart. Proxy container restarted from time to time on all of out environments, somewhere often, somewhere not. Please help identify - is it problem with Kong itself or it is issues with cloud provider.

Kong version: `3.2`.

Typical kong pods status:
```
NAME                            READY   STATUS    RESTARTS      AGE     IP             NODE                   NOMINATED NODE   READINESS GATES
gateway-kong-6476c465cb-6cwm4   2/2     Running   3 (83m ago)   4h31m   10.244.2.42    pri-production-qztnx   <none>           <none>
gateway-kong-6476c465cb-w4dtw   2/2     Running   2 (37m ago)   4h30m   10.244.3.248   pri-production-qztnr   <none>           <none>
```

Events for failed pod:
```
NAMESPACE    LAST SEEN   TYPE      REASON              OBJECT                              MESSAGE
production   23m         Warning   Unhealthy           pod/gateway-kong-6476c465cb-w4dtw   Liveness probe failed: Get "http://10.244.3.248:8100/status": context deadline exceeded (Client.Timeout exceeded while a waiting headers)
production   23m         Warning   Unhealthy           pod/gateway-kong-6476c465cb-w4dtw   Readiness probe failed: Get "http://10.244.3.248:8100/status": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
production   23m         Normal    Killing             pod/gateway-kong-6476c465cb-w4dtw   Container proxy failed liveness probe, will be restarted
production   23m         Warning   Unhealthy           pod/gateway-kong-6476c465cb-w4dtw   Readiness probe failed: Get "http://10.244.3.248:8100/status": dial tcp 10.244.3.248:8100: connect: connection refused
```

To catch this issue I `curl` livenessProbe endpoint in endless loop for both pods to see if there is errors, and actually there is NO errors until the moment when container is killed by kubelet - see attached [curl-out.txt](https://github.com/Kong/kong/files/11295369/curl-out.txt) and [log-proxy-prev-gateway-kong-6476c465cb-w4dtw.txt](https://github.com/Kong/kong/files/11295375/log-proxy-prev-gateway-kong-6476c465cb-w4dtw.txt).

Time sequence for your comfort:
* 15:57:59 - curl log begins, no errors. In previous records there is no errors too.
* 15:58:53 - event `Container proxy failed liveness probe, will be restarted`
* 15:58:53 - event `Liveness probe failed: Get "http://10.244.3.248:8100/status": context deadline exceeded (Client.Timeout exceeded while awaiting headers)`
* 15:59:09 - nginx `signal 3 (SIGQUIT) received from 1368, shutting down`
* 15:58:11 - curl until this moment - no errors!
* 15:58:11 - curl `curl: (52) Empty reply from server`
* 15:59:19 - nginx `signal 15 (SIGTERM) received from 1472, exiting`
* 15:59:19 - event `PreStopHook failed`

Curl command executed from another pod in same namespace. All 3 pods (the one where curl is executed and 2 gateways) run on different nodes:
```
kubectl -n production exec -it some-pod-0 -- bash -c "while true; do date; echo "gateway-kong-6476c465cb-6cwm4"; curl http://10.244.2.42:8100/status; echo; echo "gateway-kong-6476c465cb-w4dtw"; curl http://10.244.3.248:8100/status ; echo; done" | tee ./curl-out.txt
```

So, from perspective of k8s cluster, container is working, traffic is flowing, but it still killed - is it some kind of kubelet issue?! But at same time - we run HTTP health checks on our own software and it is never failed! If it is issue with kubelet or another part of cloud provider infrastructure other pods should fail too.

I notice often (not always!) there is bunch of calls to our app endpoint `POST /core/public/v1/users/token/refresh`. App make call to Kong API in this method. So, my bet it is co-related. I studying with dev team what this method do and how it use API. My best guess - we somehow crash proxy or API server. But still - curl do not fail, but livenessProbe do! :zany_face: 

[curl-out.txt](https://github.com/Kong/kong/files/11295369/curl-out.txt)
[desc-events.txt](https://github.com/Kong/kong/files/11295370/desc-events.txt)
[desc-pod-gateway-kong-6476c465cb-w4dtw.txt](https://github.com/Kong/kong/files/11295371/desc-pod-gateway-kong-6476c465cb-w4dtw.txt)
[get-events.txt](https://github.com/Kong/kong/files/11295372/get-events.txt)
[get-pod.txt](https://github.com/Kong/kong/files/11295373/get-pod.txt)
[log-kic-gateway-kong-6476c465cb-w4dtw.txt](https://github.com/Kong/kong/files/11295374/log-kic-gateway-kong-6476c465cb-w4dtw.txt)
[log-proxy-prev-gateway-kong-6476c465cb-w4dtw.txt](https://github.com/Kong/kong/files/11295375/log-proxy-prev-gateway-kong-6476c465cb-w4dtw.txt)

**Comment by ghost** [comment]
same with: #10001 ?

**Reply by baznikin** [reply]
Not at all! In #10001 Kong simply CPU throttling all the time having no time to answer queries. My clusters is almost idle at time, there is no user load. Resource limits not set (i.e. unlimited)

**Reply by anup-krai** [reply]
@baznikin  - Were you able to resolve this issue? I am also facing same issue.

**Reply by baznikin** [reply]
@anup-krai nope! but most likely it was our cloud provider issue. We run most of our environments on DigitalOcean.

At same time this discussion was started we fought with GitLab runner issues and it was (more or less) resolved by moving pipelines to other cloud provider. Recently we return our pipelines to DigitalOcean and... they worked nicely! for some time :D Pipelines hang occurs from time to time. I believe it is some misbehavior of their managed Kubernetes.

Issue with Kong resolved as well, I suppose they fix something on their side. There is info from all our current k8s clusters; most of them for development, most loaded 1st and last in the list. As we can see, restarts are infrequent and co-related with issued cloud provider publish online:
```
# DO, LON1
NAME                           READY   STATUS    RESTARTS     AGE
gateway-kong-5454bb654-p5nr2   2/2     Running   0            18d
gateway-kong-5454bb654-w2ndg   2/2     Running   2 (9d ago)   18d
# DO, LON1
gateway-kong-74bbc5b5d4-fg4px   2/2     Running   0          22d
gateway-kong-74bbc5b5d4-wktqk   2/2     Running   0          22d
# DO, FRA1
gateway-kong-7d87d5698b-lnfnf   2/2     Running   0             82d
gateway-kong-7d87d5698b-xj42h   2/2     Running   1 (49d ago)   82d
# DO, LON1
gateway-kong-7fc7487866-95q8f   2/2     Running   9 (12d ago)   69d
gateway-kong-7fc7487866-m8djp   2/2     Running   9 (12d ago)   69d
# DO, LON1
gateway-kong-7fc797887d-49x4s   2/2     Running   2 (4d15h ago)   4d19h
gateway-kong-7fc797887d-8k97r   2/2     Running   1 (4d15h ago)   4d19h
# DO, LON1
gateway-kong-64f89d45d9-8lhfg   2/2     Running   16 (12h ago)   18d
gateway-kong-64f89d45d9-mt7pk   2/2     Running   13 (12h ago)   18d
# DO, LON1
gateway-kong-7856856f46-frswh   2/2     Running   6 (8d ago)    29d
gateway-kong-7856856f46-zcmm9   2/2     Running   10 (8d ago)   47d
```

**Reply by anup-krai** [reply]
@baznikin - Thanks for your response. Can you please share the issues which cloud provider published so that we will also look into those areas?

**Reply by baznikin** [reply]
My provider post them on [status page](https://status.digitalocean.com/history)
